### PR TITLE
Fix: enqueue in updating secret

### LINKED ISSUE #6309: PrometheusAgent not reconciled when Secret/Configmap
### Is there an existing issue for this?

- [X] I have searched the existing issues

## Description
Prometheus Operator manages three instances: two regular Prometheus instances named "cluster1-stg-use1-kube-p-prometheus" and "kube-prom-infra-prometheus," and one Prometheus Agent instance named "kube-prom-agent-prometheus".

We use ServiceMonitors with a bearerTokenSecret, and the secret changes every 45 minutes. 

```
spec:
  endpoints:
    - bearerTokenSecret:
        key: token
        name: kube-prom-services-token
      honorLabels: true
      honorTimestamps: false
      interval: 60s
```
However, while the Operator consistently triggers a synchronization process for regular Prometheus instances upon secret changes, it does not consistently do so for Prometheus Agents. Consequently, the Prometheus Agent pod may retain the old secret in /etc/prometheus/config_out/prometheus.env.yaml. Although the sync for Prometheus Agent is occasionally triggered, such as during a restart of the Prometheus Operator pod or when a ServiceMonitor changes, it does not occur consistently with every secret change.

Upon changing the agentMode to false and redeploying, the synchronization works as expected.

## Steps to Reproduce
Set agentMode to true in values.yaml of kube-prometheus-stack
```
  prometheus:
    enabled: true
    agentMode: true
```

## Expected Result
Prometheus should synchronize once the secret is updated.

## Actual Result
The synchronization never occurs

### Prometheus Operator Version

```shell
prometheus-operator:v0.71.2 (kube-prometheus-stack v56.2.0, prometheus 2.49.1) & 
prometheus-operator:v0.69.1
```

### Kubernetes Version

```yaml
v1.27.7-eks
```

### Kubernetes Cluster Type

EKS

### How did you deploy Prometheus-Operator?

helm chart:prometheus-community/kube-prometheus-stack

### Manifests

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: PrometheusAgent
metadata:
  labels:
    app: kube-prometheus-stack-prometheus
    app.kubernetes.io/instance: kube-prom-agent-prometheus
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/part-of: kube-prometheus-stack
    app.kubernetes.io/version: 56.6.2
    chart: kube-prometheus-stack-56.6.2
    chartName: kube-prometheus-stack
    chartVersion: 56.2.0
spec:
  enableFeatures:
    - memory-snapshot-on-shutdown
  externalLabels:
    cloud_provider: aws
  externalUrl: http://kube-prom-agent-prometheus....
  hostNetwork: false
  image: >-
    ...prometheus/prometheus:v2.49.1
  imagePullSecrets:
    - name: ...
  listenLocal: false
  logFormat: logfmt
  logLevel: info
  paused: false
  podMetadata:
    labels:
      backstage.io/kubernetes-id: kube-prom-agent
      chartName: kube-prometheus-stack
      chartVersion: 56.2.0
      serviceName: kube-prom-agent
  podMonitorNamespaceSelector: {}
  podMonitorSelector:
    matchExpressions:
      - key: release
        operator: In
        values:
          - cluster1-stg-use1-kube-p-prometheus
          - kube-prom-infra-prometheus
          - kube-prom-agent-prometheus
  portName: http-web
  probeNamespaceSelector: {}
  probeSelector:
    matchLabels:
      release: kube-prom-agent-prometheus
  prometheusExternalLabelName: prometheus_agent
  remoteWrite:
...
  replicaExternalLabelName: prometheus_replica
  replicas: 1
  resources:
    limits:
      cpu: 2
      memory: 14Gi
    requests:
      cpu: 1
      memory: 8Gi
  routePrefix: /
  scrapeConfigNamespaceSelector: {}
  scrapeConfigSelector:
    matchLabels:
      release: kube-prom-agent-prometheus
  scrapeInterval: 30s
  securityContext:
    fsGroup: 2000
    runAsGroup: 2000
    runAsNonRoot: true
    runAsUser: 1000
    seccompProfile:
      type: RuntimeDefault
  serviceAccountName: kube-prom-agent-prometheus
  serviceMonitorNamespaceSelector: {}
  serviceMonitorSelector:
    matchExpressions:
      - key: release
        operator: In
        values:
          - cluster1-stg-use1-kube-p-prometheus
          - kube-prom-infra-prometheus
          - kube-prom-agent-prometheus
  shards: 1
  storage:
    volumeClaimTemplate:
      spec:
        resources:
          requests:
            storage: 10Gi
  version: v2.49.1
  walCompression: true
```

### prometheus-operator log output

```shell
{"caller":"operator.go:1074","component":"prometheusagent-controller","level":"debug","msg":"Secret updated","ts":"2024-02-15T16:00:05.336191755Z"}
{"caller":"operator.go:701","component":"prometheus-controller","level":"debug","msg":"Secret updated","ts":"2024-02-15T16:00:05.336370479Z"}
{"caller":"operator.go:403","component":"alertmanager-controller","level":"debug","msg":"Secret updated","ts":"2024-02-15T16:00:05.336418187Z"}
{"caller":"operator.go:987","component":"prometheus-controller","key":"cluster1-stg-use1-kube-prom/cluster1-stg-use1-kube-p-prometheus","level":"info","msg":"sync prometheus","ts":"2024-02-15T16:00:05.33650236Z"}
{"caller":"rules.go:188","component":"prometheus-controller","level":"debug","msg":"selected RuleNamespaces","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:05.336518897Z"}
{"caller":"rules.go:224","component":"prometheus-controller","level":"debug","msg":"selected Rules","namespace":"cluster1-stg-use1-kube-prom","prometheus":"...","ts":"2024-02-15T16:00:06.110760198Z"}
{"caller":"rules.go:100","component":"prometheus-controller","level":"debug","msg":"no PrometheusRule changes","namespace":"cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:06.203328836Z"}
{"caller":"resource_selector.go:100","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select ServiceMonitors from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:06.203445948Z"}
{"caller":"resource_selector.go:192","component":"prometheus-controller","level":"debug","msg":"selected ServiceMonitors","namespace":"cluster1-stg-use1-kube-prom","prometheus":"...","ts":"2024-02-15T16:00:06.222701921Z"}
{"caller":"resource_selector.go:356","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select PodMonitors from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:06.222750584Z"}
{"caller":"resource_selector.go:442","component":"prometheus-controller","level":"debug","msg":"selected PodMonitors","namespace":"cluster1-stg-use1-kube-prom","podmonitors":"cluster1-stg-use1-kube-prom/kube-prom-services-envoy-stats-monitor,cluster1-stg-use1-kube-prom/rt-jmx,cluster1-stg-use1-kube-prom/cluster1-stg-use1-mt-registry","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:06.223252505Z"}
{"caller":"resource_selector.go:482","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select Probes from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:06.223325159Z"}
{"caller":"resource_selector.go:589","component":"prometheus-controller","level":"debug","msg":"selected Probes","namespace":"cluster1-stg-use1-kube-prom","probes":"","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:06.223358528Z"}
{"caller":"resource_selector.go:645","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select ScrapeConfigs from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:06.223393562Z"}
{"caller":"resource_selector.go:768","component":"prometheus-controller","level":"debug","msg":"selected ScrapeConfigs","namespace":"cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","scrapeConfig":"","ts":"2024-02-15T16:00:06.223452695Z"}
{"caller":"operator.go:1398","component":"prometheus-controller","level":"debug","msg":"updating Prometheus configuration secret","ts":"2024-02-15T16:00:06.310159884Z"}
{"caller":"operator.go:403","component":"alertmanager-controller","level":"debug","msg":"Secret updated","ts":"2024-02-15T16:00:06.420237104Z"}
{"caller":"operator.go:1074","component":"prometheusagent-controller","level":"debug","msg":"Secret updated","ts":"2024-02-15T16:00:06.420263003Z"}
{"caller":"operator.go:701","component":"prometheus-controller","level":"debug","msg":"Secret updated","ts":"2024-02-15T16:00:06.420378346Z"}
{"caller":"operator.go:1024","component":"prometheus-controller","key":"cluster1-stg-use1-kube-prom/cluster1-stg-use1-kube-p-prometheus","level":"debug","msg":"reconciling statefulset","shard":"0","statefulset":"prometheus-cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:06.462904069Z"}
{"caller":"operator.go:1088","component":"prometheus-controller","key":"cluster1-stg-use1-kube-prom/cluster1-stg-use1-kube-p-prometheus","level":"debug","msg":"new statefulset generation inputs match current, skipping any actions","shard":"0","statefulset":"prometheus-cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:06.463645061Z"}
{"caller":"operator.go:987","component":"prometheus-controller","key":"cluster1-stg-use1-kube-prom/kube-prom-infra-prometheus","level":"info","msg":"sync prometheus","ts":"2024-02-15T16:00:06.463735505Z"}
{"caller":"rules.go:188","component":"prometheus-controller","level":"debug","msg":"selected RuleNamespaces","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"kube-prom-infra-prometheus","ts":"2024-02-15T16:00:06.463751335Z"}
{"caller":"rules.go:224","component":"prometheus-controller","level":"debug","msg":"selected Rules","namespace":"cluster1-stg-use1-kube-prom","prometheus":"...","ts":"2024-02-15T16:00:06.466237894Z"}
{"caller":"rules.go:100","component":"prometheus-controller","level":"debug","msg":"no PrometheusRule changes","namespace":"cluster1-stg-use1-kube-prom","prometheus":"kube-prom-infra-prometheus","ts":"2024-02-15T16:00:06.471060254Z"}
{"caller":"resource_selector.go:100","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select ServiceMonitors from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"kube-prom-infra-prometheus","ts":"2024-02-15T16:00:06.471100103Z"}
{"caller":"resource_selector.go:192","component":"prometheus-controller","level":"debug","msg":"selected ServiceMonitors","namespace":"cluster1-stg-use1-kube-prom","prometheus":"kube-prom-infra-prometheus","servicemonitors":"...","ts":"2024-02-15T16:00:06.473686318Z"}
{"caller":"resource_selector.go:356","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select PodMonitors from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"cluster1-stg-use1-kube-prom,default","prometheus":"kube-prom-infra-prometheus","ts":"2024-02-15T16:00:06.473744773Z"}
{"caller":"resource_selector.go:442","component":"prometheus-controller","level":"debug","msg":"selected PodMonitors","namespace":"cluster1-stg-use1-kube-prom","podmonitors":"","prometheus":"kube-prom-infra-prometheus","ts":"2024-02-15T16:00:06.47377467Z"}
{"caller":"resource_selector.go:482","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select Probes from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"kube-prom-infra-prometheus","ts":"2024-02-15T16:00:06.473799442Z"}
{"caller":"resource_selector.go:589","component":"prometheus-controller","level":"debug","msg":"selected Probes","namespace":"cluster1-stg-use1-kube-prom","probes":"","prometheus":"kube-prom-infra-prometheus","ts":"2024-02-15T16:00:06.473819206Z"}
{"caller":"resource_selector.go:645","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select ScrapeConfigs from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"kube-prom-infra-prometheus","ts":"2024-02-15T16:00:06.473840601Z"}
{"caller":"resource_selector.go:768","component":"prometheus-controller","level":"debug","msg":"selected ScrapeConfigs","namespace":"cluster1-stg-use1-kube-prom","prometheus":"kube-prom-infra-prometheus","scrapeConfig":"","ts":"2024-02-15T16:00:06.473861523Z"}
{"caller":"operator.go:1398","component":"prometheus-controller","level":"debug","msg":"updating Prometheus configuration secret","ts":"2024-02-15T16:00:06.51009701Z"}
{"caller":"operator.go:1024","component":"prometheus-controller","key":"cluster1-stg-use1-kube-prom/kube-prom-infra-prometheus","level":"debug","msg":"reconciling statefulset","shard":"0","statefulset":"prometheus-kube-prom-infra-prometheus","ts":"2024-02-15T16:00:06.617088236Z"}
{"caller":"operator.go:1088","component":"prometheus-controller","key":"cluster1-stg-use1-kube-prom/kube-prom-infra-prometheus","level":"debug","msg":"new statefulset generation inputs match current, skipping any actions","shard":"0","statefulset":"prometheus-kube-prom-infra-prometheus","ts":"2024-02-15T16:00:06.617808645Z"}
{"caller":"operator.go:987","component":"prometheus-controller","key":"cluster1-stg-use1-kube-prom/cluster1-stg-use1-kube-p-prometheus","level":"info","msg":"sync prometheus","ts":"2024-02-15T16:00:06.617915334Z"}
{"caller":"rules.go:188","component":"prometheus-controller","level":"debug","msg":"selected RuleNamespaces","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:06.617930796Z"}
{"caller":"rules.go:224","component":"prometheus-controller","level":"debug","msg":"selected Rules","namespace":"cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","rules":"","ts":"2024-02-15T16:00:07.407306812Z"}
{"caller":"rules.go:100","component":"prometheus-controller","level":"debug","msg":"no PrometheusRule changes","namespace":"cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:07.503440822Z"}
{"caller":"resource_selector.go:100","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select ServiceMonitors from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:07.503514625Z"}
{"caller":"resource_selector.go:192","component":"prometheus-controller","level":"debug","msg":"selected ServiceMonitors","namespace":"cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","servicemonitors":"...","ts":"2024-02-15T16:00:07.518443878Z"}
{"caller":"resource_selector.go:356","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select PodMonitors from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"cluster1-stg-use1-kube-prom,default","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:07.518514467Z"}
{"caller":"resource_selector.go:442","component":"prometheus-controller","level":"debug","msg":"selected PodMonitors","namespace":"cluster1-stg-use1-kube-prom","podmonitors":"...","ts":"2024-02-15T16:00:07.518737466Z"}
{"caller":"resource_selector.go:482","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select Probes from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:07.518762132Z"}
{"caller":"resource_selector.go:589","component":"prometheus-controller","level":"debug","msg":"selected Probes","namespace":"cluster1-stg-use1-kube-prom","probes":"","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:07.5187819Z"}
{"caller":"resource_selector.go:645","component":"prometheus-controller","level":"debug","msg":"filtering namespaces to select ScrapeConfigs from","namespace":"cluster1-stg-use1-kube-prom","namespaces":"default,cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:07.518804749Z"}
{"caller":"resource_selector.go:768","component":"prometheus-controller","level":"debug","msg":"selected ScrapeConfigs","namespace":"cluster1-stg-use1-kube-prom","prometheus":"cluster1-stg-use1-kube-p-prometheus","scrapeConfig":"","ts":"2024-02-15T16:00:07.51882362Z"}
{"caller":"operator.go:1398","component":"prometheus-controller","level":"debug","msg":"updating Prometheus configuration secret","ts":"2024-02-15T16:00:07.702110015Z"}
{"caller":"operator.go:1024","component":"prometheus-controller","key":"cluster1-stg-use1-kube-prom/cluster1-stg-use1-kube-p-prometheus","level":"debug","msg":"reconciling statefulset","shard":"0","statefulset":"prometheus-cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:07.76082425Z"}
{"caller":"operator.go:1088","component":"prometheus-controller","key":"cluster1-stg-use1-kube-prom/cluster1-stg-use1-kube-p-prometheus","level":"debug","msg":"new statefulset generation inputs match current, skipping any actions","shard":"0","statefulset":"prometheus-cluster1-stg-use1-kube-p-prometheus","ts":"2024-02-15T16:00:07.761475862Z"}
{"caller":"operator.go:471","component":"thanos-controller","cur":"2938815860","level":"debug","msg":"update handler","namespace":"cluster1-stg-use1-kube-prom","old":"2938815860","ts":"2024-02-15T16:00:30.42308439Z"}
{"caller":"operator.go:471","component":"thanos-controller","cur":"4","level":"debug","msg":"update handler","namespace":"default","old":"4","ts":"2024-02-15T16:00:30.423125827Z"}
{"caller":"operator.go:896","component":"prometheus-controller","cur":"4","level":"debug","msg":"update handler","namespace":"default","old":"4","ts":"2024-02-15T16:00:30.423112906Z"}
{"caller":"operator.go:896","component":"prometheus-controller","cur":"2938815860","level":"debug","msg":"update handler","namespace":"cluster1-stg-use1-kube-prom","old":"2938815860","ts":"2024-02-15T16:00:30.423144953Z"}
{"caller":"operator.go:567","component":"alertmanager-controller","cur":"4","level":"debug","msg":"update handler","namespace":"default","old":"4","ts":"2024-02-15T16:00:30.431632571Z"}
{"caller":"operator.go:567","component":"alertmanager-controller","cur":"2938815860","level":"debug","msg":"update handler","namespace":"cluster1-stg-use1-kube-prom","old":"2938815860","ts":"2024-02-15T16:00:30.43167534Z"}
{"caller":"resource_reconciler.go:317","component":"prometheusagent-controller","cur":"3015371717","level":"debug","msg":"update handler","old":"3015371717","resource":"statefulset","ts":"2024-02-15T16:00:30.524045769Z"}
{"caller":"resource_reconciler.go:317","component":"prometheusagent-controller","cur":"3017003755","level":"debug","msg":"update handler","old":"3017003755","resource":"statefulset","ts":"2024-02-15T16:00:30.524079305Z"}
{"caller":"resource_reconciler.go:317","component":"prometheusagent-controller","cur":"3017045605","level":"debug","msg":"update handler","old":"3017045605","resource":"statefulset","ts":"2024-02-15T16:00:30.524088436Z"}
{"caller":"operator.go:1183","component":"prometheusagent-controller","cur":"4","level":"debug","msg":"update handler","namespace":"default","old":"4","ts":"2024-02-15T16:00:30.633280142Z"}
{"caller":"operator.go:1183","component":"prometheusagent-controller","cur":"2938815860","level":"debug","msg":"update handler","namespace":"cluster1-stg-use1-kube-prom","old":"2938815860","ts":"2024-02-15T16:00:30.633319044Z"}
{"caller":"resource_reconciler.go:317","component":"prometheus-controller","cur":"3017003755","level":"debug","msg":"update handler","old":"3017003755","resource":"statefulset","ts":"2024-02-15T16:00:30.721823485Z"}
{"caller":"resource_reconciler.go:317","component":"prometheus-controller","cur":"3017045605","level":"debug","msg":"update handler","old":"3017045605","resource":"statefulset","ts":"2024-02-15T16:00:30.721865172Z"}
{"caller":"resource_reconciler.go:317","component":"prometheus-controller","cur":"3015371717","level":"debug","msg":"update handler","old":"3015371717","resource":"statefulset","ts":"2024-02-15T16:00:30.721876425Z"}
```

### Anything else?

I enabled debug mode for the operator but did not find any errors. Additionally, I found no relevant information in the operator's metrics.

### PR DESCRIPTION
**Pull request by jamshidi799**
Add `enqueueForPrometheusNamespace` to secret update handler
It will fix #6309

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Add enqueueForPrometheusNamespace in secret update handler
```

### CODE REVIEW

**Comment by haanhvu** [comment]
Thanks for the PR! Do you think we need an e2e test case to replicate the issue? 🤔

**Review by simonpasquier — APPROVED** [review]
Ideally we'd need an e2e test to verify this but it's a bit more involved. I also think that we should reduce code duplication since all controllers have similar patterns for configmap/secret handlers but again it doesn't need to be in this be in this PR.
TL;DR: I'm fine merging as-is but I'll create more issues to follow-up.

**Comment by jamshidi799** [comment]
I'm interested in working on these code duplication as well
